pandas/tests/groupby/test_categorical.py

```diff
@@ -1342,6 +1364,83 @@ def test_series_groupby_on_2_categoricals_unobserved_zeroes_or_nans(func, zero_o
         assert np.issubdtype(result.dtype, np.integer)
 
 
+def test_dataframe_groupby_on_2_categoricals_when_observed_is_true(reduction_func: str):
+    # GH 23865
+    # GH 27075
+    # Ensure that df.groupby, when 'by' is two pd.Categorical variables,
+    # does not return the categories that are not in df when observed=True
+    if reduction_func == "ngroup":
+        pytest.skip("ngroup does not return the Categories on the index")
+
+    res, unobserved_cats = _dataframe_groupby_on_2_categoricals(
+        reduction_func, observed=True
+    )
+
+    for cat in unobserved_cats:
+        assert cat not in res.index
+
+
+def _dataframe_groupby_on_2_categoricals(reduction_func: str, observed: bool):
```

Comment:
can you add some commentary here. i don't normally like having a helper function like this as hard for folks to grok, but i see might be hard to remove

Reply:
Just removed the helper function, not hard to do. Just requires repeating some of the same code across two tests.